File: assets/python/backend.py
import logging
import os
import sys

from schemas.unifmu_fmi_pb2 import (
    UnifmuHandshakeReturn
)
from schemas.unifmu_fmi_pb2 import (
    Fmi3Command,
    Fmi3StatusReturn,
    Fmi3FreeInstanceReturn,
    UnifmuFmi3SerializeReturn,
    Fmi3GetFloat32Return,
    Fmi3GetFloat64Return,
    Fmi3GetInt8Return,
    Fmi3GetUInt8Return,
    Fmi3GetInt16Return,
    Fmi3GetUInt16Return,
    Fmi3GetInt32Return,
    Fmi3GetUInt32Return,
    Fmi3GetInt64Return,
    Fmi3GetUInt64Return,
    Fmi3GetBooleanReturn, 
    Fmi3GetStringReturn,
    FmiGetBinaryReturn
)
from schemas.unifmu_fmi_pb2 import (
    Fmi2Command,
    Fmi2StatusReturn,
    Fmi2FreeInstanceReturn,
    UnifmuFmi2SerializeReturn,
    Fmi2GetRealReturn,
    Fmi2GetIntegerReturn,
    Fmi2GetBooleanReturn,
    Fmi2GetStringReturn
)

logging.basicConfig(level=logging.DEBUG)
logger = logging.getLogger(__file__)

try:
    import zmq
except ImportError:
    logger.fatal(
        "unable to import the python library 'zmq' required by the schemaless backend. "
        "please ensure that the library is present in the python environment launching the script. "
        "the missing dependencies can be installed using 'python -m pip install unifmu[python-backend]'"
    )
    sys.exit(-1)

from model import Model

def connectFmi3Model(socket):
    command = Fmi3Command
    while True:
        msg = socket.recv()
        command.parseFromString(msg)

        group = command.WhichOneof("command")
        data = getattr(command, command.WhichOneof("command"))

        if group == "Fmi3DoStep":
            result = Fmi3StatusReturn()
            result.status = model.fmi3DoStep(
                data.current_communication_point,
                data.communication_step_size,
                data.no_set_fmu_state_prior_to_current_point,
                data.event_handling_needed,
                data.terminate_simulation,
                data.early_return,
                data.last_successful_time
            )
        elif group == "Fmi3EnterInitializationMode":
            result = Fmi3StatusReturn()
            result.status = model.fmi3EnterInitializationMode(
                data.tolerance,
                data.start_time,
                data.stop_time
            )
        elif group == "Fmi3ExitInitializationMode":
            result = Fmi3StatusReturn()
            result.status = model.fmi3ExitInitializationMode()
        elif group == "Fmi3FreeInstance":
            result = Fmi3FreeInstanceReturn()
            logger.info(f"Fmi3FreeInstance received, shutting down")
            sys.exit(0)        
        elif group == "Fmi3Terminate":
            result = Fmi3StatusReturn()
            result.status = model.fmi3Terminate()
        elif group == "Fmi3Reset":
            result = Fmi3StatusReturn()
            result.status = model.fmi3Reset()
        elif group == "UnifmuSerialize":
            result = UnifmuFmi3SerializeReturn()
            (result.status, result.state) = model.unifmuFmi3Serialize()
        elif group == "UnifmuDeserialize":
            result = Fmi3StatusReturn
            result.status = model.unifmuFmi3Deserialize(
                data.state
            )
        # -- Getters --
        elif group == "Fmi3GetFloat32":
            result = Fmi3GetFloat32Return()
            (result.status, result.values) = model.fmi3GetFloat32(
                data.value_references
            )
        elif group == "Fmi3GetFloat64":
            result = Fmi3GetFloat64Return()
            (result.status, result.values) = model.fmi3GetFloat64(
                data.value_references
            )
        elif group == "Fmi3GetInt8":
            result = Fmi3GetInt8Return()
            (result.status, result.values) = model.fmi3GetInt8(
                data.value_references
            )
        elif group == "Fmi3GetUInt8":
            result = Fmi3GetUInt8Return()
            (result.status, result.values) = model.fmi3GetUInt8(
                data.value_references
            )
        elif group == "Fmi3GetInt16":
            result = Fmi3GetInt16Return()
            (result.status, result.values) = model.fmi3GetInt16(
                data.value_references
            )
        elif group == "Fmi3GetUInt16":
            result = Fmi3GetUInt16Return()
            (result.status, result.values) = model.fmi3GetUInt16(
                data.value_references
            )
        elif group == "Fmi3GetInt32":
            result = Fmi3GetInt32Return()
            (result.status, result.values) = model.fmi3GetInt32(
                data.value_references
            )
        elif group == "Fmi3GetUInt32":
            result = Fmi3GetUInt32Return()
            (result.status, result.values) = model.fmi3GetUInt32(
                data.value_references
            )
        elif group == "Fmi3GetInt64":
            result = Fmi3GetInt64Return()
            (result.status, result.values) = model.fmi3GetInt64(
                data.value_references
            )
        elif group == "Fmi3GetUInt64":
            result = Fmi3GetUInt64Return()
            (result.status, result.values) = model.fmi3GetUInt64(
                data.value_references
            )
        elif group == "Fmi3GetBoolean":
            result = Fmi3GetBooleanReturn()
            (result.status, result.values) = model.fmi3GetBoolean(
                data.value_references
            )
        elif group == "Fmi3GetString":
            result = Fmi3GetStringReturn()
            (result.status, result.values) = model.fmi3GetString(
                data.value_references
            )
        elif group == "FmiGetBinary":
            result = FmiGetBinaryReturn()
            (result.status, result.values) = model.fmi3GetBinary(
                data.value_references
            )
        # -- Setters --
        elif group == "Fmi3SetFloat32":
            result = Fmi2StatusReturn()
            result.status = model.fmi3SetFloat32(
                data.value_references,
                data.values
            )
        elif group == "Fmi3SetFloat64":
            result = Fmi2StatusReturn()
            result.status = model.fmi3SetFloat64(
                data.value_references,
                data.values
            )
        elif group == "Fmi3SetInt8":
            result = Fmi2StatusReturn()
            result.status = model.fmi3SetInt8(
                data.value_references,
                data.values
            )
        elif group == "Fmi3SetUInt8":
            result = Fmi2StatusReturn()
            result.status = model.fmi3SetUInt8(
                data.value_references,
                data.values
            )
        elif group == "Fmi3SetInt16":
            result = Fmi2StatusReturn()
            result.status = model.fmi3SetInt16(
                data.value_references,
                data.values
            )
        elif group == "Fmi3SetUInt16":
            result = Fmi2StatusReturn()
            result.status = model.fmi3SetUInt16(
                data.value_references,
                data.values
            )
        elif group == "Fmi3SetInt32":
            result = Fmi2StatusReturn()
            result.status = model.fmi3SetInt32(
                data.value_references,
                data.values
            )
        elif group == "Fmi3SetUInt32":
            result = Fmi2StatusReturn()
            result.status = model.fmi3SetUInt32(
                data.value_references,
                data.values
            )
        elif group == "Fmi3SetInt64":
            result = Fmi2StatusReturn()
            result.status = model.fmi3SetInt64(
                data.value_references,
                data.values
            )
        elif group == "Fmi3SetUInt64":
            result = Fmi2StatusReturn()
            result.status = model.fmi3SetUInt64(
                data.value_references,
                data.values
            )
        elif group == "Fmi3SetBoolean":
            result = Fmi2StatusReturn()
            result.status = model.fmi3SetBoolean(
                data.value_references,
                data.values
            )
        elif group == "Fmi3SetString":
            result = Fmi2StatusReturn()
            result.status = model.fmi3SetString(
                data.value_references,
                data.values
            )
        elif group == "Fmi3SetBinary":
            result = Fmi2StatusReturn()
            result.status = model.fmi3SetBinary(
                data.value_references,
                data.values
            )
        else:
            logger.error(f"unrecognized command '{group}' received, shutting down")
            sys.exit(-1)

        state = result.SerializeToString()
        socket.send(state)  

def connectFmi2Model(socket):
    command = Fmi2Command()

    while True:
        msg = socket.recv()
        command.ParseFromString(msg)

        group = command.WhichOneof("command")

        data = getattr(command, command.WhichOneof("command"))
        if group == "Fmi2DoStep":
            result = Fmi2StatusReturn()
            result.status = model.fmi2DoStep(
                data.current_time, 
                data.step_size, 
                data.no_step_prior
            )
        if group == "Fmi2SetDebugLogging":
            result = Fmi2StatusReturn()
            result.status = model.fmiSetDebugLogging(
                data.categores, data.logging_on
            )
        if group == "Fmi2SetupExperiment":
            result = Fmi2StatusReturn()
            result.status = model.fmi2SetupExperiment(
                data.start_time, data.stop_time, data.tolerance
            )
        elif group == "Fmi2EnterInitializationMode":
            result = Fmi2StatusReturn()
            result.status = model.fmi2EnterInitializationMode()
        elif group == "Fmi2ExitInitializationMode":
            result = Fmi2StatusReturn()
            result.status = model.fmi2ExitInitializationMode()
        elif group == "Fmi2FreeInstance":
            result = Fmi2FreeInstanceReturn()
            logger.info(f"Fmi2FreeInstance received, shutting down")
            sys.exit(0)        
        elif group == "Fmi2Terminate":
            result = Fmi2StatusReturn()
            result.status = model.fmi2Terminate()
        elif group == "Fmi2Reset":
            result = Fmi2StatusReturn()
            result.status = model.fmi2Reset()
        elif group == "UnifmuSerialize":
            result = UnifmuFmi2SerializeReturn()
            (result.status, result.state) = model.fmi2ExtSerialize()
        elif group == "Fmi2ExtDeserializeSlave":
            result = Fmi2StatusReturn()
            result.status = model.fmi2ExtDeserialize(data.state)
        # -- Getters --
        elif group == "Fmi2GetReal":
            result = Fmi2GetRealReturn()
            result.status, result.values[:] = model.fmi2GetReal(
                command.Fmi2GetReal.references
            )
        elif group == "Fmi2GetInteger":
            result = Fmi2GetIntegerReturn()
            result.status, result.values[:] = model.fmi2GetInteger(
                command.Fmi2GetInteger.references
            )
        elif group == "Fmi2GetBoolean":
            result = Fmi2GetBooleanReturn()
            result.status, result.values[:] = model.fmi2GetBoolean(
                command.Fmi2GetBoolean.references
            )
        elif group == "Fmi2GetString":
            result = Fmi2GetStringReturn()
            result.status, result.values[:] = model.fmi2GetString(
                command.Fmi2GetString.references
            )
        # -- Setters --
        elif group == "Fmi2SetReal":
            result = Fmi2StatusReturn()
            result.status = model.fmi2SetReal(
                command.Fmi2SetReal.references, command.Fmi2SetReal.values
            )
        elif group == "Fmi2SetInteger":
            result = Fmi2StatusReturn()
            result.status = model.fmi2SetInteger(
                command.Fmi2SetInteger.references, command.Fmi2SetInteger.values
            )
        elif group == "Fmi2SetBoolean":
            result = Fmi2StatusReturn()
            result.status = model.fmi2SetBoolean(
                command.Fmi2SetBoolean.references, command.Fmi2SetBoolean.values
            )
        elif group == "Fmi2SetString":
            result = Fmi2StatusReturn()
            result.status = model.fmi2SetString(
                command.Fmi2SetString.references, command.Fmi2SetString.values
            )
        else:
            logger.error(f"unrecognized command '{group}' received, shutting down")
            sys.exit(-1)

        state = result.SerializeToString()
        socket.send(state)

if __name__ == "__main__":

    model = Model()

    # initializing message queue
    context = zmq.Context()
    socket = context.socket(zmq.REQ)

    dispatcher_endpoint = os.environ["UNIFMU_DISPATCHER_ENDPOINT"]
    logger.info(f"dispatcher endpoint received: {dispatcher_endpoint}")

    socket.connect(dispatcher_endpoint)

    state = UnifmuHandshakeReturn().SerializeToString()
    socket.send(state)

    connectFmi2Model(socket)
    connectFmi3Model(socket)

File: assets/csharp/fmi3/backend.cs

using System.IO;
using System;
using Fmi3Messages;
using UnifmuHandshake;
using System.Collections.Generic;
using NetMQ.Sockets;
using Google.Protobuf;
using NetMQ;
using System;
using System.Diagnostics;
using System.Linq;

namespace Launch
{
    class Program
    {


        public static void Main(string[] args)
        {
            Trace.Listeners.Add(new ConsoleTraceListener()); // Logs to console
            var references_to_attr = new Dictionary<uint, string>();
            Model model = null;

            string dispatcher_endpoint = System.Environment.GetEnvironmentVariable("UNIFMU_DISPATCHER_ENDPOINT");
            if (dispatcher_endpoint == null)
            {
                Console.Error.WriteLine("Environment variable 'UNIFMU_DISPATCHER_ENDPOINT' is not set in the current enviornment.");
                Environment.Exit(-1);
            }


            var socket = new RequestSocket();
            socket.Connect(dispatcher_endpoint);


            IMessage message = new HandshakeReply{
                Status = HandshakeStatus.Ok
            };


            socket.SendFrame(message.ToByteArray(), false);
            Fmi3Command command;

            while (true)
            {
                command = Fmi3Command.Parser.ParseFrom(socket.ReceiveFrameBytes());
                Trace.TraceInformation("Command: " + command);

                switch (command.CommandCase)
                {

                    case Fmi3Command.CommandOneofCase.Fmi3InstantiateCoSimulation:
                        {
                            model = new Model(
                                command.Fmi3InstantiateCoSimulation.InstanceName,
                                command.Fmi3InstantiateCoSimulation.InstantiationToken,
                                command.Fmi3InstantiateCoSimulation.ResourcePath,
                                command.Fmi3InstantiateCoSimulation.Visible,
                                command.Fmi3InstantiateCoSimulation.LoggingOn,
                                command.Fmi3InstantiateCoSimulation.EventModeUsed,
                                command.Fmi3InstantiateCoSimulation.EarlyReturnAllowed,
                                command.Fmi3InstantiateCoSimulation.RequiredIntermediateVariables.ToList()
                            );
                            var result = new Fmi3EmptyReturn();
                            message = result;
                        }
                        break;

                    case Fmi3Command.CommandOneofCase.Fmi3EnterInitializationMode:
                        {
                            var result = new Fmi3StatusReturn();
                            result.Status = model.Fmi3EnterInitializationMode();
                            message = result;
                        }

                        break;

                    case Fmi3Command.CommandOneofCase.Fmi3ExitInitializationMode:
                        {
                            var result = new Fmi3StatusReturn();
                            result.Status = model.Fmi3ExitInitializationMode();
                            message = result;
                        }
                        break;

                    case Fmi3Command.CommandOneofCase.Fmi3EnterConfigurationMode:
                        {
                            var result = new Fmi3StatusReturn();
                            result.Status = model.Fmi3EnterConfigurationMode();
                            message = result;
                        }

                        break;

                    case Fmi3Command.CommandOneofCase.Fmi3ExitConfigurationMode:
                        {
                            var result = new Fmi3StatusReturn();
                            result.Status = model.Fmi3ExitConfigurationMode();
                            message = result;
                        }
                        break;

                    case Fmi3Command.CommandOneofCase.Fmi3EnterEventMode:
                        {
                            var result = new Fmi3StatusReturn();
                            result.Status = model.Fmi3EnterEventMode();
                            message = result;
                        }

                        break;

                    case Fmi3Command.CommandOneofCase.Fmi3EnterStepMode:
                        {
                            var result = new Fmi3StatusReturn();
                            result.Status = model.Fmi3EnterStepMode();
                            message = result;
                        }

                        break;

                    case Fmi3Command.CommandOneofCase.Fmi3DoStep:
                        {
                            var result = new Fmi3DoStepReturn();
                            (var status, var event_handling_needed, var terminate_simulation, var early_return, var last_successful_time) = model.Fmi3DoStep(command.Fmi3DoStep.CurrentCommunicationPoint, command.Fmi3DoStep.CommunicationStepSize, command.Fmi3DoStep.NoSetFmuStatePriorToCurrentPoint);
                            result.Status = status;
                            result.EventHandlingNeeded = event_handling_needed;
                            result.TerminateSimulation = terminate_simulation;
                            result.EarlyReturn = early_return;
                            result.LastSuccessfulTime = last_successful_time;
                            message = result;
                        }

                        break;

                    case Fmi3Command.CommandOneofCase.Fmi3UpdateDiscreteStates:
                        {
                            var result = new Fmi3UpdateDiscreteStatesReturn();
                            (var status, var discrete_states_need_update, var terminate_simulation, var nominals_continuous_states_changed, var values_continuous_states_changed, var next_event_time_defined, var next_event_time) = model.Fmi3UpdateDiscreteStates();
                            result.Status = status;
                            result.DiscreteStatesNeedUpdate = discrete_states_need_update;
                            result.TerminateSimulation = terminate_simulation;
                            result.NominalsContinuousStatesChanged = nominals_continuous_states_changed;
                            result.ValuesContinuousStatesChanged = values_continuous_states_changed;
                            result.NextEventTimeDefined = next_event_time_defined;
                            result.NextEventTime = next_event_time;
                            message = result;
                        }

                        break;

                    case Fmi3Command.CommandOneofCase.Fmi3SetFloat32:
                        {
                            var result = new Fmi3StatusReturn();
                            result.Status = model.Fmi3SetFloat32(command.Fmi3SetFloat32.ValueReferences, command.Fmi3SetFloat32.Values);
                            message = result;
                        }
                        break;

                    case Fmi3Command.CommandOneofCase.Fmi3SetFloat64:
                        {
                            var result = new Fmi3StatusReturn();
                            result.Status = model.Fmi3SetFloat64(command.Fmi3SetFloat64.ValueReferences, command.Fmi3SetFloat64.Values);
                            message = result;
                        }
                        break;


                    case Fmi3Command.CommandOneofCase.Fmi3SetInt8:
                        {
                            var result = new Fmi3StatusReturn();
                            result.Status = model.Fmi3SetInt8(
                                command.Fmi3SetInt8.ValueReferences,
                                command.Fmi3SetInt8.Values);
                            message = result;

                        }
                        break;

                    case Fmi3Command.CommandOneofCase.Fmi3SetUInt8:
                        {
                            var result = new Fmi3StatusReturn();
                            result.Status = model.Fmi3SetUInt8(
                                command.Fmi3SetUInt8.ValueReferences,
                                command.Fmi3SetUInt8.Values);
                            message = result;

                        }
                        break;

                    case Fmi3Command.CommandOneofCase.Fmi3SetInt16:
                        {
                            var result = new Fmi3StatusReturn();
                            result.Status = model.Fmi3SetInt16(
                                command.Fmi3SetInt16.ValueReferences,
                                command.Fmi3SetInt16.Values);
                            message = result;

                        }
                        break;

                    case Fmi3Command.CommandOneofCase.Fmi3SetUInt16:
                        {
                            var result = new Fmi3StatusReturn();
                            result.Status = model.Fmi3SetUInt16(
                                command.Fmi3SetUInt16.ValueReferences,
                                command.Fmi3SetUInt16.Values);
                            message = result;

                        }
                        break;

                    case Fmi3Command.CommandOneofCase.Fmi3SetInt32:
                        {
                            var result = new Fmi3StatusReturn();
                            result.Status = model.Fmi3SetInt32(
                                command.Fmi3SetInt32.ValueReferences,
                                command.Fmi3SetInt32.Values);
                            message = result;

                        }
                        break;

                    case Fmi3Command.CommandOneofCase.Fmi3SetUInt32:
                        {
                            var result = new Fmi3StatusReturn();
                            result.Status = model.Fmi3SetUInt32(
                                command.Fmi3SetUInt32.ValueReferences,
                                command.Fmi3SetUInt32.Values);
                            message = result;

                        }
                        break;

                    case Fmi3Command.CommandOneofCase.Fmi3SetInt64:
                        {
                            var result = new Fmi3StatusReturn();
                            result.Status = model.Fmi3SetInt64(
                                command.Fmi3SetInt64.ValueReferences,
                                command.Fmi3SetInt64.Values);
                            message = result;

                        }
                        break;

                    case Fmi3Command.CommandOneofCase.Fmi3SetUInt64:
                        {
                            var result = new Fmi3StatusReturn();
                            result.Status = model.Fmi3SetUInt64(
                                command.Fmi3SetUInt64.ValueReferences,
                                command.Fmi3SetUInt64.Values);
                            message = result;

                        }
                        break;

                    case Fmi3Command.CommandOneofCase.Fmi3SetBoolean:
                        {
                            var result = new Fmi3StatusReturn();
                            result.Status = model.Fmi3SetBoolean(
                                command.Fmi3SetBoolean.ValueReferences,
                                command.Fmi3SetBoolean.Values
                            );
                            message = result;


                        }
                        break;

                    case Fmi3Command.CommandOneofCase.Fmi3SetBinary:
                        {
                            var result = new Fmi3StatusReturn();
                            result.Status = model.Fmi3SetBinary(
                                command.Fmi3SetBinary.ValueReferences,
                                command.Fmi3SetBinary.ValueSizes,
                                ConvertToByteArrayList(command.Fmi3SetBinary.Values)
                            );
                            message = result;


                        }
                        break;

                    case Fmi3Command.CommandOneofCase.Fmi3SetClock:
                        {
                            var result = new Fmi3StatusReturn();
                            result.Status = model.Fmi3SetClock(
                                command.Fmi3SetClock.ValueReferences,
                                command.Fmi3SetClock.Values
                            );
                            message = result;


                        }
                        break;

                    case Fmi3Command.CommandOneofCase.Fmi3SetIntervalDecimal:
                        {
                            var c = command.Fmi3SetIntervalDecimal;
                            var res = model.Fmi3SetIntervalDecimal(c.ValueReferences, c.Intervals);
                            var result = new Fmi3StatusReturn
                            {
                                Status = res,
                            };
                            message = result;
                        }
                        break;


                    case Fmi3Command.CommandOneofCase.Fmi3SetIntervalFraction:
                        {
                            var c = command.Fmi3SetIntervalFraction;
                            var res = model.Fmi3SetIntervalFraction(c.ValueReferences, c.Counters, c.Resolutions);
                            var result = new Fmi3StatusReturn
                            {
                                Status = res,
                            };
                            message = result;
                        }
                        break;

                    case Fmi3Command.CommandOneofCase.Fmi3SetShiftDecimal:
                        {
                            var c = command.Fmi3SetShiftDecimal;
                            var res = model.Fmi3SetShiftDecimal(c.ValueReferences, c.Shifts);
                            var result = new Fmi3StatusReturn
                            {
                                Status = res,
                            };
                            message = result;
                        }
                        break;

                    case Fmi3Command.CommandOneofCase.Fmi3SetShiftFraction:
                        {
                            var c = command.Fmi3SetShiftFraction;
                            var res = model.Fmi3SetShiftFraction(c.ValueReferences, c.Counters, c.Resolutions);
                            var result = new Fmi3StatusReturn
                            {
                                Status = res,
                            };
                            message = result;                        
                        }
                        break;


                    case Fmi3Command.CommandOneofCase.Fmi3SetString:
                        {
                            var result = new Fmi3StatusReturn();
                            result.Status = model.Fmi3SetString(command.Fmi3SetString.ValueReferences, command.Fmi3SetString.Values);
                            message = result;
                        }
                        break;


                    case Fmi3Command.CommandOneofCase.Fmi3GetFloat32:
                        {
                            var result = new Fmi3GetFloat32Return();
                            (var status, var values) = model.Fmi3GetFloat32(command.Fmi3GetFloat32.ValueReferences);
                            result.Values.AddRange(values);
                            result.Status = status;
                            message = result;
                        }
                        break;

                    case Fmi3Command.CommandOneofCase.Fmi3GetFloat64:
                        {
                            var result = new Fmi3GetFloat64Return();
                            (var status, var values) = model.Fmi3GetFloat64(command.Fmi3GetFloat64.ValueReferences);
                            result.Values.AddRange(values);
                            result.Status = status;
                            message = result;
                        }
                        break;

                    case Fmi3Command.CommandOneofCase.Fmi3GetInt8:
                        {
                            var result = new Fmi3GetInt8Return();
                            (var status, var values) = model.Fmi3GetInt8(command.Fmi3GetInt8.ValueReferences);
                            result.Values.AddRange(values);
                            result.Status = status;
                            message = result;

                        }
                        break;

                    case Fmi3Command.CommandOneofCase.Fmi3GetUInt8:
                        {
                            var result = new Fmi3GetUInt8Return();
                            (var status, var values) = model.Fmi3GetUInt8(command.Fmi3GetUInt8.ValueReferences);
                            result.Values.AddRange(values);
                            result.Status = status;
                            message = result;

                        }
                        break;

                    case Fmi3Command.CommandOneofCase.Fmi3GetInt16:
                        {
                            var result = new Fmi3GetInt16Return();
                            (var status, var values) = model.Fmi3GetInt16(command.Fmi3GetInt16.ValueReferences);
                            result.Values.AddRange(values);
                            result.Status = status;
                            message = result;

                        }
                        break;

                    case Fmi3Command.CommandOneofCase.Fmi3GetUInt16:
                        {
                            var result = new Fmi3GetUInt16Return();
                            (var status, var values) = model.Fmi3GetUInt16(command.Fmi3GetUInt16.ValueReferences);
                            result.Values.AddRange(values);
                            result.Status = status;
                            message = result;

                        }
                        break;

                    case Fmi3Command.CommandOneofCase.Fmi3GetInt32:
                        {
                            var result = new Fmi3GetInt32Return();
                            (var status, var values) = model.Fmi3GetInt32(command.Fmi3GetInt32.ValueReferences);
                            result.Values.AddRange(values);
                            result.Status = status;
                            message = result;

                        }
                        break;

                    case Fmi3Command.CommandOneofCase.Fmi3GetUInt32:
                        {
                            var result = new Fmi3GetUInt32Return();
                            (var status, var values) = model.Fmi3GetUInt32(command.Fmi3GetUInt32.ValueReferences);
                            result.Values.AddRange(values);
                            result.Status = status;
                            message = result;

                        }
                        break;

                    case Fmi3Command.CommandOneofCase.Fmi3GetInt64:
                        {
                            var result = new Fmi3GetInt64Return();
                            (var status, var values) = model.Fmi3GetInt64(command.Fmi3GetInt64.ValueReferences);
                            result.Values.AddRange(values);
                            result.Status = status;
                            message = result;

                        }
                        break;

                    case Fmi3Command.CommandOneofCase.Fmi3GetUInt64:
                        {
                            var result = new Fmi3GetUInt64Return();
                            (var status, var values) = model.Fmi3GetUInt64(command.Fmi3GetUInt64.ValueReferences);
                            result.Values.AddRange(values);
                            result.Status = status;
                            message = result;

                        }
                        break;

                    case Fmi3Command.CommandOneofCase.Fmi3GetBoolean:
                        {
                            var result = new Fmi3GetBooleanReturn();
                            (var status, var values) = model.Fmi3GetBoolean(command.Fmi3GetBoolean.ValueReferences);
                            result.Values.AddRange(values);
                            result.Status = status;
                            message = result;

                        }
                        break;

                    case Fmi3Command.CommandOneofCase.Fmi3GetBinary:
                        {
                            var result = new Fmi3GetBinaryReturn();
                            (var status, var values) = model.Fmi3GetBinary(command.Fmi3GetBinary.ValueReferences);
                            result.Values.AddRange(ConvertToByteStringList(values));
                            result.Status = status;
                            message = result;

                        }
                        break;

                    case Fmi3Command.CommandOneofCase.Fmi3GetClock:
                        {
                            var result = new Fmi3GetClockReturn();
                            (var status, var values) = model.Fmi3GetClock(command.Fmi3GetClock.ValueReferences);
                            result.Values.AddRange(values);
                            result.Status = status;
                            message = result;

                        }
                        break;

                    case Fmi3Command.CommandOneofCase.Fmi3GetString:
                        {
                            var result = new Fmi3GetStringReturn();
                            (var status, var values) = model.Fmi3GetString(command.Fmi3GetString.ValueReferences);
                            result.Values.AddRange(values);
                            result.Status = status;
                            message = result;

                        }
                        break;

                    case Fmi3Command.CommandOneofCase.Fmi3GetIntervalDecimal:
                        {
                            var c = command.Fmi3GetIntervalDecimal;
                            (var status, var intervals, var qualifiers) = model.Fmi3GetIntervalDecimal(c.ValueReferences);
                            var result = new Fmi3GetIntervalDecimalReturn();
                            result.Status = status;
                            result.Intervals.AddRange(intervals);
                            result.Qualifiers.AddRange(qualifiers);
                            message = result;
                        }
                        break;

                    case Fmi3Command.CommandOneofCase.Fmi3GetIntervalFraction:
                        {
                            var c = command.Fmi3GetIntervalFraction;
                            (var status, var counters, var resolutions, var qualifiers) = model.Fmi3GetIntervalFraction(c.ValueReferences);
                            var result = new Fmi3GetIntervalFractionReturn();
                            result.Status = status;
                            result.Counters.AddRange(counters);
                            result.Resolutions.AddRange(resolutions);
                            result.Qualifiers.AddRange(qualifiers);
                            message = result;
                        }
                        break;

                    case Fmi3Command.CommandOneofCase.Fmi3GetShiftDecimal:
                        {
                            var c = command.Fmi3GetShiftDecimal;
                            (var status, var shifts) = model.Fmi3GetShiftDecimal(c.ValueReferences);
                            var result = new Fmi3GetShiftDecimalReturn();
                            result.Status = status;
                            result.Shifts.AddRange(shifts);
                            message = result;
                        }
                        break;

                    case Fmi3Command.CommandOneofCase.Fmi3GetShiftFraction:
                        {
                            var c = command.Fmi3GetShiftFraction;
                            (var status, var counters, var resolutions) = model.Fmi3GetShiftFraction(c.ValueReferences);
                            var result = new Fmi3GetShiftFractionReturn();
                            result.Status = status;
                            result.Counters.AddRange(counters);
                            result.Resolutions.AddRange(resolutions);
                            message = result;
                        }
                        break;


                    case Fmi3Command.CommandOneofCase.Fmi3Reset:
                        {
                            var result = new Fmi3StatusReturn();
                            result.Status = model.Fmi3Reset();
                            message = result;

                        }

                        break;

                    case Fmi3Command.CommandOneofCase.Fmi3Terminate:
                        {
                            var result = new Fmi3StatusReturn();
                            result.Status = model.Fmi3Terminate();
                            message = result;
                        }
                        break;

                    case Fmi3Command.CommandOneofCase.Fmi3SerializeFmuState:
                        {
                            var result = new Fmi3SerializeFmuStateReturn();
                            (var status, var state) = model.Fmi3SerializeFmuState();
                            result.Status = status;
                            result.State = ByteString.CopyFrom(state);
                            message = result;
                        }
                        break;
                    case Fmi3Command.CommandOneofCase.Fmi3DeserializeFmuState:
                        {
                            var result = new Fmi3StatusReturn();
                            result.Status = model.Fmi3DeserializeFmuState(command.Fmi3DeserializeFmuState.State.ToByteArray());
                            message = result;
                        }

                        break;

                    case Fmi3Command.CommandOneofCase.Fmi3FreeInstance:
                        {
                            Console.WriteLine("received fmi3FreeInstance, exiting with status code 0");
                            Environment.Exit(0);
                        }


                        break;

                    default:
                        Console.Error.WriteLine("unrecognized command {0}, exiting with status code -1", command.CommandCase);
                        Environment.Exit(-1);
                        break;
                }


                socket.SendFrame(message.ToByteArray(), false);
            }

        }

        public static IEnumerable<byte[]> ConvertToByteArrayList(IEnumerable<ByteString> byteStrings)
        {
            var byteArrayList = new List<byte[]>();
            foreach (var bs in byteStrings)
            {
                byteArrayList.Add(bs.ToByteArray());
            }
            return byteArrayList;
        }

        public static IEnumerable<ByteString> ConvertToByteStringList(IEnumerable<byte[]> byteArrays)
        {
            var byteStringList = new List<ByteString>();
            foreach (var arr in byteArrays)
            {
                byteStringList.Add(ByteString.CopyFrom(arr));
            }
            return byteStringList;
        }
    }
}